File: brady/linked_list_cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

// O(n) time, O(1) space
bool hasCycle(struct ListNode *head) {
    if (head == NULL) return false;

    struct ListNode *slow = head;
    struct ListNode *fast = head->next;

    // If fast exhausts the list, then there is no cycle. If slow == fast then there is a cycle
    while (fast != NULL && fast->next != NULL) {
        if (slow == fast) return true;
        slow = slow->next;
        fast = fast->next->next;
    }

    return false;
}

File: sharjeel/linkedListCycle.py
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        
        if head is None or head.next is None: return False

        # Two Pointer Solution
        slow = head
        fast = head.next.next

        while fast and fast.next:
            if fast == slow and slow: return True
            fast = fast.next.next
            slow = slow.next

        return False